source/extensions/filters/http/on_demand/on_demand_update.cc

```diff
@@ -3,17 +3,53 @@
 #include "source/common/common/assert.h"
 #include "source/common/common/enum_to_int.h"
 #include "source/common/common/logger.h"
+#include "source/common/config/xds_resource.h"
 #include "source/common/http/codes.h"
+#include "source/common/http/utility.h"
+#include "source/extensions/filters/http/well_known_names.h"
 
 namespace Envoy {
 namespace Extensions {
 namespace HttpFilters {
 namespace OnDemand {
 
+namespace {
+
+template <typename ProtoConfig>
+Upstream::OdCdsApiHandlePtr createOdCdsApi(const ProtoConfig& proto_config,
+                                           Upstream::ClusterManager& cm,
+                                           ProtobufMessage::ValidationVisitor& validation_visitor) {
+  if (!proto_config.has_odcds_config() && proto_config.resources_locator().empty()) {
+    return nullptr;
```

Comment:
If possible, we should aim to avoid returning null, as we shift the handling to the caller of this method

Reply:
The null pointer means that the on-demand cluster discovery is disabled and that's handled in `OnDemandRouteUpdate::handleOnDemandCDS`. Avoiding returning a nullptr here would mean I'd need to return some behavior object like:

```c++
struct DecodeHeadersBehavior {
    virtual Http::FilterHeadersStatus decodeHeaders() = 0;
};

struct RDSDecodeHeadersBehavior {
    Http::FilterHeadersStatus decodeHeaders() override {
        // only request a route config update if route is missing 
    }
};

struct RDSCDSDecodeHeadersBehavior {
    Http::FilterHeadersStatus decodeHeaders() override {
        // request a route config update if route is missing
        // request a cluster if its missing
    }
};

std::unique_ptr<DecodeHeadersBehavior> getBehavior(const ProtoConfig& proto_config) {
    if (!proto_config.has_odcds_config() && …) {
        return std::make_unique<RDSDecodeHeadersBehavior> (…);
    }
    return std::make_unique<RDSCDSDecodeHeadersBehavior> (…);
}
```

Would a thing like this be ok?

Reply:
Rewritten using the behavior classes.